library/std/src/path/tests.rs

```diff
@@ -1394,6 +1394,46 @@ pub fn test_set_extension() {
     tfe!("/", "foo", "/", false);
 }
 
+#[test]
+pub fn test_with_extension() {
+    macro_rules! twe (
+        ($input:expr, $extension:expr, $expected:expr) => ({
+            let input = Path::new($input);
+            let output = input.with_extension($extension);
+
+            assert!(
+                output.to_str() == Some($expected),
+                "calling Path::new({:?}).with_extension({:?}): Expected {:?}, got {:?}",
+                $input, $extension, $expected, output,
+            );
+        });
+    );
+
+    twe!("foo", "txt", "foo.txt");
+    twe!("foo.bar", "txt", "foo.txt");
+    twe!("foo.bar.baz", "txt", "foo.bar.txt");
+    twe!(".test", "txt", ".test.txt");
+    twe!("foo.txt", "", "foo");
+    twe!("foo", "", "foo");
+    twe!("", "foo", "");
+    twe!(".", "foo", ".");
+    twe!("foo/", "bar", "foo.bar");
+    twe!("foo/.", "bar", "foo.bar");
+    twe!("..", "foo", "..");
+    twe!("foo/..", "bar", "foo/..");
+    twe!("/", "foo", "/");
```

Comment:
After inspecting these, I realized that my implementation over-allocates in some of the corner cases.

```rust
/* ok     */ twe!("foo", "txt", "foo.txt");
/* ok     */ twe!("foo.bar", "txt", "foo.txt");
/* ok     */ twe!("foo.bar.baz", "txt", "foo.bar.txt");
/* ok     */ twe!(".test", "txt", ".test.txt");
/* over 1 */ twe!("foo.txt", "", "foo");
/* over 1 */ twe!("foo", "", "foo");
/* over 4 */ twe!("", "foo", "");
/* over 4 */ twe!(".", "foo", ".");
/* over 1 */ twe!("foo/", "bar", "foo.bar");
/* over 1 */ twe!("foo/.", "bar", "foo.bar");
/* over 4 */ twe!("..", "foo", "..");
/* over 4 */ twe!("foo/..", "bar", "foo/..");
/* over 4 */ twe!("/", "foo", "/");
```

For this one specifically, it does an extra (potentially) unused allocation, because it's an empty path:
```rust
/* over 4 */ twe!("", "foo", "");
 ```